Fix GitHub Action scripts to push all nuget packages

Shane32 opened the pull request:
Fixes #1936

Shane32 commented:
Before:
![image](https://user-images.githubusercontent.com/6377684/96965433-24eefc00-14da-11eb-803d-1866960a6e53.png)


After:
![image](https://user-images.githubusercontent.com/6377684/96965380-0db00e80-14da-11eb-82ff-0af43dfadc78.png)

sungam3r commented:
Maybe try `out/*.nupkg` instead of `out/*` first ? If it works, the script will look simpler.

sungam3r commented:
Or just `*.nupkg` if the current dir is already `src/out`.

sungam3r commented:
Also you can try republish release with `--skip-duplicate` switch. I think this will require removing the tag and re-adding it.

Shane32 commented:
> Or just `*.nupkg` if the current dir is already `src/out`.

I'm sure it will work.

Shane32 commented:
That is, I'm sure it will work for the new syntax with the for loop.

sungam3r commented:
I still suggest trying without a loop. Perhaps it's all about the search or relative path.

sungam3r commented:
You see that physically the files are there.

Shane32 commented:
It doesn't work

![image](https://user-images.githubusercontent.com/6377684/97009272-e5440680-1511-11eb-88a0-1005f50e004d.png)

Shane32 commented:
> You see that physically the files are there.

Yes, and they upload as artifacts in the prior step.

It's possible that the wildcard syntax only works on a Windows OS; just guessing though.

Shane32 commented:
I patched the current release
![image](https://user-images.githubusercontent.com/6377684/97010289-30aae480-1513-11eb-9eaf-939933ffa0a1.png)

sungam3r commented:
> It's possible that the wildcard syntax only works on a Windows OS; just guessing though.

Hmm, maybe doublequotes?

![изображение](https://user-images.githubusercontent.com/21261007/97015577-66bc8880-1554-11eb-8130-54c9b2d1595c.png)

> I patched the current release.

How? This PR is not yet in the master.

sungam3r commented:
Here it is https://github.com/NuGet/Home/issues/4393

sungam3r commented:
https://github.com/NuGet/Home/issues/4393#issuecomment-660504976

you can work around the issue by running dotnet push '*.nupkg'. By putting the * in single quotes, it tells the shell not to glob and pass directly to the program unmodified, so that it mimicks the two shells available on Windows by default, and runs NuGet's internal file globbing.

sungam3r commented:
![Uploading изображение.png…]()

sungam3r commented:
Wow. This was a very interesting issue.

Shane32 commented:
> How? This PR is not yet in the master.

I created a branch and modified the workflow file to run on any push for that branch name, with necessary parameter to skip duplicates.  I also had to modify the workflow to set `ref=3.1.0` because of course there's no tag associated with that branch.  Then I deleted the branch.

You can view the workflow file and the run here:
https://github.com/graphql-dotnet/graphql-dotnet/actions/runs/324232823

Shane32 commented:
> running dotnet push '*.nupkg'. By putting the * in single quotes, it tells the shell not to glob and pass directly to the program unmodified

Cool.  I'll try it.

Shane32 commented:
Works!!!!!!!!!  @sungam3r = 🥇 
![image](https://user-images.githubusercontent.com/6377684/97025109-a15afc80-1525-11eb-92ad-9700595c435b.png)

sungam3r commented:
Oh, those quotes!

Shane32 commented on .github/workflows/publish-code.yml:
```diff
@@ -70,5 +70,9 @@ jobs:
-        working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.NUGET_AUTH_TOKEN}}
+        working-directory: src/out
+        run: |
+          for file in *
+          do
+            dotnet nuget push "$file" -k ${{secrets.NUGET_AUTH_TOKEN}}
+          done
```
```suggestion
        run: dotnet nuget push "*.nupkg" -k ${{secrets.NUGET_AUTH_TOKEN}}
```

Shane32 commented:
Should be all set @sungam3r 

Nice work finding that issue and fix

sungam3r commented on .github/workflows/build-artifacts-code.yml:
```diff
@@ -62,5 +62,5 @@ jobs:
       - name: Publish Nuget packages to GitHub registry
-        working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.GITHUB_TOKEN}}
+        working-directory: src/out
```
So, now you can even keep `src` as current dir if you like. Then the current dir will be the same throughout the script.

Shane32 replied:
Yeah; it shouldn't matter but I'll change it back; hold on

Shane32 replied:
I just hadn't tested it this way.

Shane32 replied:
Done

Shane32 commented on .github/workflows/build-artifacts-code.yml:
```diff
@@ -62,5 +62,5 @@ jobs:
       - name: Publish Nuget packages to GitHub registry
-        working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.GITHUB_TOKEN}}
+        working-directory: src/out
+        run: dotnet nuget push "*.nupkg" -k ${{secrets.GITHUB_TOKEN}}
```
```suggestion
        working-directory: src
        run: dotnet nuget push "out/*.nupkg" -k ${{secrets.GITHUB_TOKEN}}
```

Shane32 commented on .github/workflows/publish-code.yml:
```diff
@@ -70,5 +70,5 @@ jobs:
       - name: Publish Nuget packages to Nuget registry
-        working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.NUGET_AUTH_TOKEN}}
+        working-directory: src/out
+        run: dotnet nuget push "*.nupkg" -k ${{secrets.NUGET_AUTH_TOKEN}}
```
```suggestion
        working-directory: src
        run: dotnet nuget push "out/*.nupkg" -k ${{secrets.NUGET_AUTH_TOKEN}}
```

Shane32 commented:
> you can work around the issue by running dotnet push '*.nupkg'. By putting the * in single quotes, it tells the shell not to glob and pass directly to the program unmodified, so that it mimicks the two shells available on Windows by default, and runs NuGet's internal file globbing.

The whole concept of shell meddling with the arguments seems absurd to me.  But I'm sure to a linux person the opposite would be true.

sungam3r commented on .github/workflows/build-artifacts-code.yml:
```diff
@@ -63,4 +63,4 @@ jobs:
       - name: Publish Nuget packages to GitHub registry
         working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.GITHUB_TOKEN}}
+        run: dotnet nuget push "out/*.nupkg" -k ${{secrets.GITHUB_TOKEN}}
```
```suggestion
        run: dotnet nuget push "out/*" -k ${{secrets.GITHUB_TOKEN}}
```

This may sound funny, but we can now even remove the file extension.

sungam3r replied:
Do as you see fit. I like to see extension explicitly.

sungam3r commented on .github/workflows/publish-code.yml:
```diff
@@ -71,4 +71,4 @@ jobs:
       - name: Publish Nuget packages to Nuget registry
         working-directory: src
-        run: dotnet nuget push out/* -k ${{secrets.NUGET_AUTH_TOKEN}}
+        run: dotnet nuget push "out/*.nupkg" -k ${{secrets.NUGET_AUTH_TOKEN}}
```
```suggestion
        run: dotnet nuget push "out/*" -k ${{secrets.NUGET_AUTH_TOKEN}}
```

sungam3r reviewed (approved):
Long adventure and all because of the quotes.
